array.c

```diff
@@ -6706,31 +6706,40 @@ push_value(st_data_t key, st_data_t val, st_data_t ary)
 
 /*
  *  call-seq:
- *     ary.uniq!                -> ary or nil
- *     ary.uniq! {|item| ...}   -> ary or nil
+ *    array.uniq! -> self or nil
+ *    array.uniq! {|element| ... } -> self or nil
  *
- *  Removes duplicate elements from +self+.
+ *  Removes duplicate elements from +self+, the first occurrence always being retained;
+ *  returns +self+ if any elements removed, +nil+ otherwise.
  *
- *  If a block is given, it will use the return value of the block for
- *  comparison.
+ *  ---
  *
- *  It compares values using their #hash and #eql? methods for efficiency.
+ *  With no block given, identifies and removes elements using method <tt>eql?</tt>
+ *  to compare.
  *
- *  +self+ is traversed in order, and the first occurrence is kept.
+ *  Returns +self+ if any elements removed:
+ *    a = [0, 0, 1, 1, 2, 2]
+ *    a1 = a.uniq!
+ *    a1 # => [0, 1, 2]
+ *    a1.equal?(a) # => true # Returned self
  *
- *  Returns +nil+ if no changes are made (that is, no duplicates are found).
+ *  Returns +nil+ if no elements removed:
+ *    [0, 1, 2].uniq! # => nil
  *
- *     a = [ "a", "a", "b", "b", "c" ]
- *     a.uniq!   # => ["a", "b", "c"]
+ *  ---
  *
- *     b = [ "a", "b", "c" ]
- *     b.uniq!   # => nil
+ *  With a block given, calls the block for each element;
+ *  removes elements for which the block returns duplicate values.
```

Comment:
Is the duplicate value check with a block using `eql?` or `==` (I assume the former, but haven't checked)?  It should probably be documented considering you are documenting that without a block, it uses `eql?`.  Same for uniq below.

Reply:
Don't know how to determine this.

Reply:
It's got to be `eql?` and `hash` otherwise it would be `O(n^2)`...

Reply:
A way to test it (without referring to the source):
```ruby
class S; def eql?(x); false; end; def ==(x); true; end; end
[S.new, S.new].uniq.size # => 1
```

Reply:
Here's a test program showing that `eql?` is used in both cases:

```ruby
class A
  def eql?(x) p [:eql?, self, x]; true end
  def ==(x) p ['==', self, :x]; false end
  def hash; p [:hash, self]; 1 end
  p [new, new].uniq
  p [new, new].uniq{|x| x}
end
```

Output:

```
[:hash, #<A:0x00000096dc4e7a78>]
[:hash, #<A:0x00000096dc4e7a50>]
[:eql?, #<A:0x00000096dc4e7a50>, #<A:0x00000096dc4e7a78>]
[#<A:0x00000096dc4e7a78>]
[:hash, #<A:0x00000096dc4e6bf0>]
[:hash, #<A:0x00000096dc4e6bc8>]
[:eql?, #<A:0x00000096dc4e6bc8>, #<A:0x00000096dc4e6bf0>]
[#<A:0x00000096dc4e6bf0>]
```